provider/clickhouse.go

```diff
@@ -996,20 +1047,88 @@ func (store *clickHouseOfflineStore) GetTrainingSet(id ResourceID) (TrainingSetI
 	for _, name := range columnNames {
 		features = append(features, sanitizeCH(name.Name))
 	}
-	columns := strings.Join(features[:], ", ")
-	trainingSetQry := store.query.trainingRowSelect(columns, trainingSetName)
+	columns := strings.Join(features, ", ")
+
+	return &TrainingSetPreparation{
+		TrainingSetName: trainingSetName,
+		Columns:         columns,
+	}, nil
+}
+
+func (store *clickHouseOfflineStore) GetTrainingSet(id ResourceID) (TrainingSetIterator, error) {
+	fmt.Printf("Getting Training Set: %v\n", id)
+	prep, err := store.prepareTrainingSetQuery(id)
+	if err != nil {
+		return nil, err
+	}
+	trainingSetQry := store.query.trainingRowSelect(prep.Columns, prep.TrainingSetName)
 	fmt.Printf("Training Set Query: %s\n", trainingSetQry)
 	rows, err := store.db.Query(trainingSetQry)
 	if err != nil {
 		return nil, fferr.NewResourceExecutionError(pt.ClickHouseOffline.String(), id.Name, id.Variant, fferr.ResourceType(id.Type.String()), err)
 	}
-	colTypes, err := store.getValueColumnTypes(trainingSetName)
+	colTypes, err := store.getValueColumnTypes(prep.TrainingSetName)
 	if err != nil {
 		return nil, err
 	}
 	return store.newsqlTrainingSetIterator(rows, colTypes), nil
 }
 
+func (store *clickHouseOfflineStore) GetTrainingSetTestSplit(
+	id ResourceID,
+	testSize float32,
+	shuffle bool,
+	randomState int,
+) (TrainingSetIterator, TrainingSetIterator, func() error, error) {
+	prep, err := store.prepareTrainingSetQuery(id)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	trainingTestSplitName, err := store.CreateTrainingTestSplit(
+		prep.TrainingSetName,
+		testSize,
+		shuffle,
+		randomState,
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	train, test := store.query.trainingRowSplitSelect(prep.Columns, sanitizeCH(trainingTestSplitName))
+	fmt.Printf("Training Set Query: %s\n", train)
+	fmt.Printf("Test Set Query: %s\n", test)
+	testRows, err := store.db.Query(test)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("could not query test set: %v", err)
+
+	}
+	trainRows, err := store.db.Query(train)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("could not query train set: %v", err)
+
+	}
+
+	colTypes, err := store.getValueColumnTypes(trainingTestSplitName)
+	fmt.Printf("these are the column types: %v\n", colTypes)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("could not get column types: %v", err)
```

Comment:
If this happened, dropFunc would never be called right?

Maybe force the caller to call create and defer close then get the iterators. Would also make the function signature a little cleaner.